crates/gpio/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+use std::collections::hash_map::HashMap;
+use std::sync::Arc;
+
+use implementors::*;
+
+use slight_common::{impl_resource, BasicState};
+use slight_file::capability_store::CapabilityStore;
+use slight_file::resource::GpioResource::*;
+use slight_file::Resource;
+
+mod implementors;
+#[cfg(test)]
+mod tests;
+
+wit_bindgen_wasmtime::export!("../../wit/gpio.wit");
+wit_error_rs::impl_error!(gpio::GpioError);
+
+/// Implements the GPIO interface defined by gpio.wit.
+///
+/// This structure is responsible for constructing the pin resources described in the slightfile and providing them to the application upon request.
+///
+/// It must be [Send], [Sync], and [Clone].
+#[derive(Clone)]
+pub struct Gpio {
+    pins: HashMap<String, Result<Pin, gpio::GpioError>>,
+}
+
+/// A type for storing constructed pin resources.
+///
+/// There should be one variant for each pin type, holding an [Arc] reference to the implementor trait object.
+#[derive(Debug, Clone)]
+enum Pin {
+    Input(Arc<dyn InputPinImplementor>),
+    Output(Arc<dyn OutputPinImplementor>),
+    PwmOutput(Arc<dyn PwmOutputPinImplementor>),
+}
+
+impl Gpio {
+    /// Construct a new [Gpio] object.
+    ///
+    /// This function reads in the pin descriptors from the named state in `capability_store`.
+    pub fn new(name: &str, capability_store: CapabilityStore<BasicState>) -> Self {
+        let state = capability_store.get(name, "gpio").unwrap().clone();
+        let mut implementor: Box<dyn GpioImplementor> =
+            match GpioImplementors::from(state.implementor) {
+                #[cfg(feature = "raspberry_pi")]
+                GpioImplementors::RaspberryPi => Box::new(raspberry_pi::PiGpioImplementor::new()),
+            };
+
+        let pins = state
+            .configs_map
+            .map(|configs| {
+                configs
+                    .iter()
+                    .map(|(name, config)| (name.clone(), implementor.parse_pin_config(config)))
+                    .collect()
+            })
+            .unwrap_or_else(HashMap::new);
```

Comment:
Currently, all pin usage is registered via the slightfile – Is there any usage to registering pins are runtime?